presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java

```diff
@@ -445,4 +399,229 @@ private static TypeSignatureParameter applyBoundVariables(TypeSignatureParameter
+                if (typeSignatureParameter.getKind() == ParameterKind.VARIABLE) {
+                    Long typeParameter = bindings.getLongVariable(typeSignatureParameter.getVariable());
+                    if (typeParameter == null) {
+                        // if an existing value doesn't exist for the given variable name, use the value that comes from the actual type.
+                        Optional<Type> type = typeManager.coerceTypeBase(actualType, formalTypeSignature.getBase());
+                        if (!type.isPresent()) {
+                            return SolverReturnStatus.UNSOLVABLE;
+                        }
+                        TypeSignature typeSignature = type.get().getTypeSignature();
+                        originalTypeTypeParametersBuilder.add(TypeSignatureParameter.of(typeSignature.getParameters().get(i).getLongLiteral()));
+                    }
+                    else {
+                        originalTypeTypeParametersBuilder.add(TypeSignatureParameter.of(typeParameter));
+                    }
+                }
+                else {
+                    verify(typeSignatureParameter.getKind() == ParameterKind.LONG);
+                    originalTypeTypeParametersBuilder.add(typeSignatureParameter);
+                }
+            }
+            Type originalType = typeManager.getType(new TypeSignature(formalTypeSignature.getBase(), originalTypeTypeParametersBuilder.build()));
+            Optional<Type> commonSuperType = typeManager.getCommonSuperType(originalType, actualType);
+            if (!commonSuperType.isPresent()) {
+                return SolverReturnStatus.UNSOLVABLE;
+            }
+            TypeSignature commonSuperTypeSignature = commonSuperType.get().getTypeSignature();
+            if (!commonSuperTypeSignature.getBase().equals(formalTypeSignature.getBase())) {
+                return SolverReturnStatus.UNSOLVABLE;
+            }
+            SolverReturnStatus result = SolverReturnStatus.UNCHANGED;
+            for (int i = 0; i < parameters.size(); i++) {
+                TypeSignatureParameter typeSignatureParameter = parameters.get(i);
+                long commonSuperLongLiteral = commonSuperTypeSignature.getParameters().get(i).getLongLiteral();
+                if (typeSignatureParameter.getKind() == ParameterKind.VARIABLE) {
+                    String variableName = typeSignatureParameter.getVariable();
+                    Long typeParameter = bindings.getLongVariable(variableName);
+                    if (typeParameter == null || typeParameter != commonSuperLongLiteral) {
+                        bindings.setLongVariable(variableName, commonSuperLongLiteral);
+                        result = SolverReturnStatus.CHANGED;
+                    }
+                }
+                else {
+                    verify(typeSignatureParameter.getKind() == ParameterKind.LONG);
+                    if (commonSuperLongLiteral != typeSignatureParameter.getLongLiteral()) {
```

Comment:
This relies on the assumption that all literal parameters are upper bounds. That is, for all T with literal parameters, and x < y, `T(x)` is a subtype of `T(y)`. Probably worth putting in a comment mentioning that